### PR TITLE
adding a geometry module

### PR DESCRIPTION
**Pull request by meg-1**
### Describe your change:
implementing suggestion: [specifics] (https://github.com/TheAlgorithms/Python/pull/8778#issuecomment-1600934020) which was brought up in #8778 

This is a geometry module dedicated to working with different geometrical structures and shapes in a strucutred manner.

**The main interfaces are:**
1. Shape - an interface which represents all possible geometrical shapes
2. Shape Types (submodule) - Interfaces for different types of shapes
3. Geometrical Primitives - Utility Data Structures for working with shapes

Feel free to extend this module.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by cclauss** [comment]
A lot of these errors could be automatically fixed with:
```
python3 -m pip install ruff
ruff --select=I,TID --fix geometry
```

**Comment by cclauss** [comment]
OK...  I created a new file that contains a bunch of feedback on how to structure things and how to create doctests.  The current pull request contains tons of files which makes it difficult to review but there is great stuff here.  Rectangles need four Sides, not two.  Using dataclasses makes building doctests easier.

**Comment by meg-1** [comment]
> OK... I created a new file that contains a bunch of feedback on how to structure things and how to create doctests. The current pull request contains tons of files which makes it difficult to review but there is great stuff here. Rectangles need four Sides, not two. Using dataclasses makes building doctests easier.

@cclauss , Thank you incredibly for the thorough code review, it is grately appreciated! I have a few clarifying questions, just to check whether I understood everything correctly, if you don't mind, to not waste more of your time in the future:

1. Should we keep all of the code in one file, just as the code review showed, or was the code placed in one file simply for a more efficient code review, and the original structure is fine?

2. Do you think it would be a good idea to move all of the primitives (Angle, Side) to a file (f.e. 'primities.py'), and Shapes into a different file (f.e. 'shapes.py'), as a more structured approach (leave files such as open_shape, closed_shape, e.t.c.)?

3. Based on 'The Algorithms' structure, and the mission of the repo, would it be better to leave the structure as it is right now, complicated yet easy to add onto in the future (f.e. adding a new class to the structure), if anyone wishes, or to simplify it by modifying the existing code directly (f.e. adding a new method to an existing class), yet making it harder to add onto in the future?

File: geometry/code_review_feedback.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
+from __future__ import annotations
+import math
```
Comment (cclauss) [inline geometry/code_review_feedback.py:2]:
```suggestion

import math
```

File: geometry/code_review_feedback.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
+from __future__ import annotations
+
+import math
+from dataclasses import dataclass, field
+from types import NoneType
+from typing import Self
+
+
```
Comment (cclauss) [inline geometry/code_review_feedback.py:8]:
```suggestion
```

File: geometry/geometry.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+import math
```
Comment (cclauss) [inline geometry/geometry.py:2]:
```suggestion

import math
```

**Comment by cclauss** [comment]
```
codespell................................................................Failed
- hook id: codespell
- exit code: 65

geometry/geometry.py:43: lightyears ==> light years
geometry/code_review_feedback.py:43: lightyears ==> light years
```
Hint: This should not be `light years` anyway.  Put something better instead.

File: geometry/code_review_feedback.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    50
+    """
+
+    def __init__(self, short_side_length: float, long_side_length: float) -> None:
+        super().__init__()
+        self.short_side_length = short_side_length
+        self.long_side_length = long_side_length
+        self.post_init()
+
+    def post_init(self):
```
Comment (cclauss) [inline geometry/code_review_feedback.py:218]:
```suggestion
    def post_init(self) -> None:
```

File: geometry/code_review_feedback.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        >>> Rectangle(5, 10)  # doctest: +NORMALIZE_WHITESPACE
+        Rectangle(sides=[Side(length=5, angle=Angle(degrees=90), next_side=None),
+                         Side(length=10, angle=Angle(degrees=90), next_side=None)])
+        """
+        self.short_side = Side(self.short_side_length)
+        self.long_side = Side(self.long_side_length)
+        super().add_side(self.short_side)
+        super().add_side(self.long_side)
+
+    def perimeter(self):
```
Comment (cclauss) [inline geometry/code_review_feedback.py:229]:
```suggestion
    def perimeter(self) -> float:
```

File: geometry/code_review_feedback.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        return (self.short_side.length + self.long_side.length) * 2
+
+    def area(self):
```
Comment (cclauss) [inline geometry/code_review_feedback.py:232]:
```suggestion
    def area(self) -> float:
```

File: geometry/geometry.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        >>> Rectangle(5, 10)  # doctest: +NORMALIZE_WHITESPACE
+        Rectangle(sides=[Side(length=5, angle=Angle(degrees=90), next_side=None),
+        Side(length=10, angle=Angle(degrees=90), next_side=None)])
+        """
+        self.short_side = Side(self.short_side_length)
+        self.long_side = Side(self.long_side_length)
+        super().add_side(self.short_side)
+        super().add_side(self.long_side)
+
+    def perimeter(self):
```
Comment (cclauss) [inline geometry/geometry.py:229]:
```suggestion
    def perimeter(self) -> float:
```

File: geometry/geometry.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        return (self.short_side.length + self.long_side.length) * 2
+
+    def area(self):
```
Comment (cclauss) [inline geometry/geometry.py:232]:
```suggestion
    def area(self) -> float:
```

File: geometry/geometry.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    a structure which represents a
+    geometrical square on a 2D surface
+    >>> square_one = Square(5)
+    >>> square_one.perimeter()
+    20
+    >>> square_one.area()
+    25
+    """
+
+    def __init__(self, side_length):
```
Comment (cclauss) [inline geometry/geometry.py:248]:
```suggestion
    def __init__(self, side_length) -> None:
```

File: geometry/geometry.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        super().__init__(side_length, side_length)
+
+    def perimeter(self):
```
Comment (cclauss) [inline geometry/geometry.py:251]:
```suggestion
    def perimeter(self) -> float:
```

File: geometry/geometry.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        return super().perimeter()
+
+    def area(self):
```
Comment (cclauss) [inline geometry/geometry.py:254]:
```suggestion
    def area(self) -> float:
```

File: geometry/geometry.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    50
+    """
+
+    def __init__(self, short_side_length: float, long_side_length: float) -> None:
+        super().__init__()
+        self.short_side_length = short_side_length
+        self.long_side_length = long_side_length
+        self.post_init()
+
+    def post_init(self):
```
Comment (cclauss) [inline geometry/geometry.py:218]:
```suggestion
    def post_init(self) -> None:
```

**Review by cclauss — CHANGES_REQUESTED** [review]

File: geometry/geometry.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+    a structure which represents a
+    geometrical square on a 2D surface
+    >>> square_one = Square(5)
+    >>> square_one.perimeter()
+    20
+    >>> square_one.area()
+    25
+    """
+
+    def __init__(self, side_length) -> None:
```
Comment (cclauss) [inline geometry/geometry.py:248]:
```suggestion
    def __init__(self, side_length: float) -> None:
```

**Review by cclauss — CHANGES_REQUESTED** [review]

File: geometry/geometry.py [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+            raise TypeError("degrees must be a numeric value between 0 and 360.")
+
+
+@dataclass
+class Side:
+    """
+    A side of a two dimensional Shape such as Polygon, etc.
+    adjacent_sides: a list of sides which are adjacent to the current side
+    angle: the angle in degrees between each adjacent side
+    length: the length of the current side in lightyears
```
Comment (cclauss) [inline geometry/geometry.py:43]:
```suggestion
    length: the length of the current side in meters
```

**Review by cclauss — APPROVED** [review]

**Comment by cclauss** [comment]
Thanks for doing this.  Sorry that it took so long to land.

In a future pull request(s) could be broken into the following files.
* `angle.py`
* `side.py`
* `ellipse.py` includes Ellipse and Circle (Perhaps `Oval` should be the base class ;-) )
* `polygon.py` includes Rectangle, Square, (future...) Trapezoid, Triangle, Pentagon, Hexigon, etc...

* `Circle` has optimized calculations for `area` but `Square` does not yet have such optimizations.
* `Rectangle` has two sides, not four!

**Pull request merged**